sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -550,6 +565,220 @@ case class FilterEstimation(plan: Filter, catalystConf: CatalystConf) extends Lo
+      attrLeft: Attribute,
+      attrRight: Attribute,
+      update: Boolean): Option[Double] = {
+
+    if (!colStatsMap.contains(attrLeft)) {
+      logDebug("[CBO] No statistics for " + attrLeft)
+      return None
+    }
+    if (!colStatsMap.contains(attrRight)) {
+      logDebug("[CBO] No statistics for " + attrRight)
+      return None
+    }
+
+    attrLeft.dataType match {
+      case StringType | BinaryType =>
+        // TODO: It is difficult to support other binary comparisons for String/Binary
+        // type without min/max and advanced statistics like histogram.
+        logDebug("[CBO] No range comparison statistics for String/Binary type " + attrLeft)
+        return None
+      case _ =>
+    }
+
+    val colStatLeft = colStatsMap(attrLeft)
+    val statsRangeLeft = Range(colStatLeft.min, colStatLeft.max, attrLeft.dataType)
+      .asInstanceOf[NumericRange]
+    val maxLeft = BigDecimal(statsRangeLeft.max)
+    val minLeft = BigDecimal(statsRangeLeft.min)
+
+    val colStatRight = colStatsMap(attrRight)
+    val statsRangeRight = Range(colStatRight.min, colStatRight.max, attrRight.dataType)
+      .asInstanceOf[NumericRange]
+    val maxRight = BigDecimal(statsRangeRight.max)
+    val minRight = BigDecimal(statsRangeRight.min)
+
+    // determine the overlapping degree between predicate range and column's range
+    val (noOverlap: Boolean, completeOverlap: Boolean) = op match {
+      // Left < Right or Left <= Right
+      // - no overlap:
+      //      minRight           maxRight     minLeft       maxLeft
+      // 0 ------+------------------+------------+-------------+------->
+      // - complete overlap:
+      //      minLeft            maxLeft      minRight      maxRight
+      // 0 ------+------------------+------------+-------------+------->
+      case _: LessThan =>
+        (minLeft >= maxRight, maxLeft < minRight)
+      case _: LessThanOrEqual =>
+        (minLeft > maxRight, maxLeft <= minRight)
+
+      // Left > Right or Left >= Right
+      // - no overlap:
+      //      minLeft            maxLeft      minRight      maxRight
+      // 0 ------+------------------+------------+-------------+------->
+      // - complete overlap:
+      //      minRight           maxRight     minLeft       maxLeft
+      // 0 ------+------------------+------------+-------------+------->
+      case _: GreaterThan =>
+        (maxLeft <= minRight, minLeft > maxRight)
+      case _: GreaterThanOrEqual =>
+        (maxLeft < minRight, minLeft >= maxRight)
+
+      // Left = Right or Left <=> Right
+      // - no overlap:
+      //      minLeft            maxLeft      minRight      maxRight
+      // 0 ------+------------------+------------+-------------+------->
+      //      minRight           maxRight     minLeft       maxLeft
+      // 0 ------+------------------+------------+-------------+------->
+      // - complete overlap:
+      //      minLeft            maxLeft
+      //      minRight           maxRight
```

Comment:
I think `Left = Right` is different from the other 2 cases, even the range completely overlaps, the filter selectivity is not 100%.